tests/Agency/StorePerformanceTest.cpp

```diff
@@ -0,0 +1,423 @@
+  write(std::vector<std::vector<std::string>>{{to_json_object(obj)}});
+  auto const json_read_key {std::string("[[\"") + key + "\"]]"};
+  for (int i{}; i < hammer_times; ++i) {
+    obj[key] = std::to_string(i);
+    auto result {read(json_read_key)};
+  }
+}
+
+// write lots of different keys in different places
+//  Random-access performance
+
+static std::string rand_path(const size_t len = 10u) {
+  static auto const alphabet {"abcdefghijklmnopqrstuvwxyz0123456789"};
+  std::string result;
+  std::generate_n(std::back_inserter(result), 1 + std::rand() % (len-1), [](){
+    return alphabet[std::rand() % (sizeof(alphabet)-1)];
+  });
+  return result;
+}
+
+#include <iostream>
+
+TEST_F(StorePerformanceTest, scattered_keys_w) {
+  for (int i{}; i < scattered_times; ++i) {
+    auto const key {rand_path()};
+    auto const json_object {"{\"" + key + "\": " + std::to_string(rand()) + "}"};
+    auto result {write(std::vector<std::vector<std::string>>{{json_object}})};
+    ASSERT_EQ(result.front(), consensus::apply_ret_t::APPLIED);
+  }
+}
+
+TEST_F(StorePerformanceTest, scattered_keys_wr) {
+  std::vector<std::string> keys;
+  std::generate_n(std::back_inserter(keys), scattered_times, [](){ return rand_path(); });
+  for (auto const& key: keys) {
+    writeAndCheck("[[{\"" + key + "\":1}]]");
+  }
+  for (int i{}; i < scattered_times; ++i) {
+    auto const& key {keys[std::rand() % keys.size()]};
+    auto result {read("[[\"" + key + "\"]]")};
+    ASSERT_EQ(result->toJson(), "[{\"" + key + "\":1}]");
+  }
+}
+
+TEST_F(StorePerformanceTest, scattered_keys_wwr) {
+  std::vector<std::string> keys;
+  std::generate_n(std::back_inserter(keys), scattered_times, [](){ return rand_path(20); });
+  for (auto const& key: keys) {
+    auto const json_object {"{\"" + key + "\": 1}"};
+    auto result {write(std::vector<std::vector<std::string>>{{json_object}})};
+    ASSERT_EQ(result.front(), consensus::apply_ret_t::APPLIED);
+  }
+  for (auto const& key: keys) {
+    auto const json_object {"[[{\"" + key + "\": {\"op\": \"increment\"}}]]"};
+    auto result {write(json_object)};
+    if (result.front() != consensus::apply_ret_t::APPLIED) {
+      throw std::runtime_error(json_object + " could not be applied: " + std::to_string(result.front()));
+    }
+  }
+  for (int i{}; i < scattered_times; ++i) {
+    auto const& key {keys[std::rand() % keys.size()]};
+    auto result {read("[[\"" + key + "\"]]")};
+    ASSERT_EQ(result->toJson(), "[{\"" + key + "\":2}]");
+  }
+}
+
+// do a lot of small transactions
+// Performance in situations that should smartly use caches and primary memory.
+TEST_F(StorePerformanceTest, small_tx_r) {
+  writeAndCheck("[[{\"a\": 1}]]");
+  writeAndCheck("[[{\"b/b/c\": 2}]]");
+  writeAndCheck("[[{\"d\": 3}]]");
+  for (int i{}; i < hammer_times; ++i) {
+    auto const result {read("[[\"a\"], [\"b/b/c\"], [\"d\"]]")};
+    ASSERT_EQ(result->toJson(), "[{\"a\":1},{\"b\":{\"b\":{\"c\":2}}},{\"d\":3}]");
```

Comment:
We should not assert in a tight loop which is supposed to measure the Store performance.

Reply:
I thought about that, but then I also thought that a smart implementation could eventually delay some operations until the result is actually used, in which case the test would become invalid. What are your thoughts?

Reply:
Sorry to re-iterate, what are your thoughts about this?

Reply:
I guess this could be true, and the assert is probably not too expensive, let's leave it as it is.